File: backend/app/schemas/prescription.py
"""Prescription request/response schemas."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class MedicationItem(BaseModel):
    """A single medication in a prescription."""

    name: str = Field(..., min_length=1, max_length=200)
    dosis: str = Field(..., min_length=1, max_length=100)
    frecuencia: str = Field(..., min_length=1, max_length=100)
    duracion_dias: int = Field(..., ge=1, le=365)
    via: str = Field(default="oral", max_length=30)
    instrucciones: str | None = None


class PrescriptionCreate(BaseModel):
    """Fields required to create a new prescription."""

    medications: list[MedicationItem] = Field(..., min_length=1, max_length=20)
    diagnosis_id: str | None = None
    notes: str | None = None


class PrescriptionResponse(BaseModel):
    """Full prescription detail."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    patient_id: str
    doctor_id: str
    medications: list[dict]
    diagnosis_id: str | None = None
    notes: str | None = None
    is_active: bool
    created_at: datetime
    updated_at: datetime


class PrescriptionListResponse(BaseModel):
    """Paginated list of prescriptions."""

    items: list[PrescriptionResponse]
    total: int
    page: int
    page_size: int

File: frontend/lib/validations/prescription.ts
/**
 * Zod validation schemas for prescription forms.
 *
 * These mirror the backend Pydantic schemas for prescription creation.
 * All field names use snake_case to match the backend API.
 */

import { z } from "zod";

// ─── Via (Route of Administration) ────────────────────────────────────────────

export const VIA_OPTIONS = [
  "oral",
  "sublingual",
  "topica",
  "intramuscular",
  "intravenosa",
  "rectal",
] as const;

export type Via = (typeof VIA_OPTIONS)[number];

export const VIA_LABELS: Record<string, string> = {
  oral: "Oral",
  sublingual: "Sublingual",
  topica: "Tópica",
  intramuscular: "Intramuscular",
  intravenosa: "Intravenosa",
  rectal: "Rectal",
};

// ─── Medication Item Schema ────────────────────────────────────────────────────

/**
 * Schema for a single medication entry within a prescription.
 */
export const medicationItemSchema = z.object({
  name: z
    .string()
    .trim()
    .min(1, "Nombre requerido")
    .max(200, "El nombre no puede exceder 200 caracteres"),

  dosis: z
    .string()
    .trim()
    .min(1, "Dosis requerida")
    .max(100, "La dosis no puede exceder 100 caracteres"),

  frecuencia: z
    .string()
    .trim()
    .min(1, "Frecuencia requerida")
    .max(100, "La frecuencia no puede exceder 100 caracteres"),

  duracion_dias: z.coerce
    .number({ invalid_type_error: "Ingresa un número de días válido" })
    .int("La duración debe ser un número entero")
    .min(1, "Mínimo 1 día")
    .max(365, "Máximo 365 días"),

  via: z.string().max(30, "Vía de administración inválida").default("oral"),

  instrucciones: z
    .string()
    .max(500, "Las instrucciones no pueden exceder 500 caracteres")
    .optional()
    .nullable()
    .transform((v) => v?.trim() || null),
});

export interface MedicationItemFormValues {
  name: string;
  dosis: string;
  frecuencia: string;
  duracion_dias: number;
  via: string;
  instrucciones: string | null;
}

// ─── Prescription Create Schema ───────────────────────────────────────────────

/**
 * Schema for creating a new prescription.
 * Mirrors the backend PrescriptionCreate Pydantic schema.
 */
export const prescriptionCreateSchema = z.object({
  medications: z
    .array(medicationItemSchema)
    .min(1, "Agregue al menos un medicamento")
    .max(20, "Se permiten máximo 20 medicamentos"),

  diagnosis_id: z
    .string()
    .uuid("ID de diagnóstico inválido")
    .optional()
    .nullable(),

  notes: z
    .string()
    .max(2000, "Las notas no pueden exceder 2000 caracteres")
    .optional()
    .nullable()
    .transform((v) => v?.trim() || null),
});

export type PrescriptionCreate = z.infer<typeof prescriptionCreateSchema>;
